src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs

```diff
@@ -632,9 +635,18 @@ private async Task ProcessRequests<TContext>(IHttpApplication<TContext> applicat
                     // Run the application code for this request
                     await application.ProcessRequestAsync(context);
 
-                    if (!_connectionAborted)
+                    // Trigger OnStarting if it hasn't been called yet and the app hasn't
+                    // already failed. If an OnStarting callback throws we can go through
+                    // our normal error handling in ProduceEnd.
+                    // https://github.com/aspnet/KestrelHttpServer/issues/43
+                    if (!HasResponseStarted && _applicationException == null && _onStarting?.Count > 0)
                     {
-                        VerifyResponseContentLength();
+                        await FireOnStarting();
```

Comment:
FireOnStarting was moved to prevent users from modifying ContentLength after the check.